File: Leetcode/Python3/424. Longest Repeating Character Replacement/solution.py
# https://leetcode.com/problems/longest-repeating-character-replacement/


from collections import Counter


class Solution:  # Optimized
    def characterReplacement(self, s: str, k: int) -> int:
        left = 0
        used_alpha = {}
        max_len = 0

        for right in range(len(s)):
            if s[right] not in used_alpha:
                used_alpha[s[right]] = 1
            else:
                used_alpha[s[right]] += 1

            length = right - left + 1
            if max(length - used_alpha.value()) <= k:
                max_len = max(max_len, length)
            else:
                used_alpha[s[left]] -= 1
                if used_alpha[s[left]] == 0:
                    del used_alpha[s[left]]
                left += 1

        return max_len


class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        left = 0
        count = Counter()
        for right in range(len(s)):
            count[s[right]] += 1

            most = count.most_common(1)[0][1]
            if (right - left) - most == k:
                result = max(result, right - left)
            elif (right - left) - most > k:
                count[s[left]] -= 1
                left += 1
        return right - left

File: Leetcode/C++/424. Longest Repeating Character Replacement/solution.cpp
// https://leetcode.com/problems/longest-repeating-character-replacement/description/

#include <string>
#include <map>

using namespace std;

class Solution
{
public:
    int characterReplacement(string s, int k)
    {
        int longest = 0, left = 0;
        unordered_map<char, int> counter;
        for (int right = 0; right < s.size(); right++)
        {
            counter[s[right]] += 1;
            int mostChar = mostCommon(counter);
            if (mostChar + k < right - left + 1) // Most common char cant cover left, rigth range with replacement move left index
            {
                counter[s[left]] -= 1;
                left += 1;
            }
            longest = max(longest, right - left + 1);
        }
        return longest;
    }

    int mostCommon(unordered_map<char, int> &counter) // Find mostCommon char in left, right range
    {
        int maxCount = INT_MIN;
        for (auto &pair : counter)
        {
            if (maxCount < pair.second)
            {
                maxCount = pair.second;
            }
        }
        return maxCount;
    }
};